docs/Simulation/Software-RTL-Simulation.rst

```diff
@@ -138,3 +138,16 @@ An open-source vcd-capable waveform viewer is `GTKWave <http://gtkwave.sourcefor
 For a VCS simulation, this will generate a vpd file (this is a proprietary waveform representation format used by Synopsys) that can be loaded to vpd-supported waveform viewers.
 If you have Synopsys licenses, we recommend using the DVE waveform viewer.
 
+.. _sw-sim-verilator-opts:
+
+Additional Verilator Options
+-------------------------------
+When building the verilator simulator there are several additional options:
+
+.. code-block:: shell
+
+   make VL8_THREADS=8 ENABLE_PRINTF_PATTERN='<python-regex>'
+
+The ``VL8_THREADS=<num>`` option enables the compiled verilator simulator to use ``<num>`` parallel threads. on a multi-socket machine, you will want to make sure all threads are on the same socket by using ``numactl``. You can also just use the ``numa_prefix`` wrapper, which is a simple wrapper around ``numactl`` that runs your verilated simulator like this: ``$(numa_prefix) ./simulator-<name> <simulator-args>``.
```

Comment:
So, is the user supposed to do this themselves? IMO this should all be handled in the `Makefile` so that you can do:
```
make run-binary BINARY=foo.riscv VL8_THREADS=8
```

Reply:
no the simulator is hard-coded to use the specified number of threads. `VL8_THREADS` is only used during compilation

Reply:
My point still stands, except that variable would be used when building the simulator instead of when running the executable.

Reply:
i think the current makefile already works this way?

you just run `make VERILATOR_THREADS=10 <config_name>` to build the simulator

Reply:
but `$(numa_prefix)` is not in the `run-binary` or `run` `common.mk` targets, unless I'm missing something. Or can you do `$(numa_prefix) make run`?

Reply:
i don't have a good answer for this. i usually just run the simulator manually like `./simulator-... <args>`, so i can just run `$(numa_prefix) ./simulator-... <args>`. 

if we put `$(numa_prefix)` in the Makefiles, we'd have to check to make sure `numactl` was installed and the user actually wants to use it...

Reply:
I'd guess you're probably in the minority running the simulator outside of make (since there are a lot of args to pass). Is `numactl` strictly necessary or just a perfomance optimization? It's also easy to have `variables.mk` check for existence and set a sane default (or "") for `$(numa_prefix)`, if desired.

Reply:
Out of curiosity, how large is the penalty from not setting `numa_prefix`?

Reply:
@jerryz123 2x performance improvement when using 20 threads on a 2-socket xeon with 20 cores per socket

Reply:
the verilator multithreading manual says its essentially required: https://www.veripool.org/projects/verilator/wiki/Manual-verilator#MULTITHREADING

Reply:
on a 1-socket machine it probably makes little difference

Reply:
I think it would be prudent to work this into the build system, then. We probably want a make variable that sets this simulation prefix (`numa_prefix` is probably fine, although we usually upcase all these variables), then put it into `common.mk` and set the variable to empty for VCS. Probably also need some kind of guard to prevent building a multithreaded simulator if the machine lacks `numactl`

Reply:
`numa_prefix` is an executable command actually, not a variable. but sure, we can do `NUMA_PREFIX=$(shell numa_prefix)`. 

btw, `numa_prefix` always works and can be used in front of any command, including `echo`. it will work on a 1-socket or multi-socket machine. i even use it on my laptop. so i see no actual need to apply numa_ctrl selectively based on the machine's properties

Reply:
Sure, as long as we can set `NUMA_PREFIX=` for VCS (or if there is some other reason the user doesn't want it), that should be fine

Reply:
fixed